doc/man3/SSL_shutdown.pod

```diff
@@ -21,161 +21,242 @@ SSL_shutdown, SSL_shutdown_ex - shut down a TLS/SSL or QUIC connection
 
 =head1 QUIC-SPECIFIC SHUTDOWN CONSIDERATIONS
 
-When using QUIC, SSL_shutdown() or SSL_shutdown_ex() causes any data written to
-a stream which has not yet been sent to the peer to be written before the
-shutdown process is considered complete. An exception to this is streams which
-terminated in a non-normal fashion, for example due to a stream reset; only
-streams which are non-terminated or which terminated in a normal fashion have
-their pending send buffers flushed in this manner. This behaviour can be skipped
-by setting the B<SSL_SHUTDOWN_FLAG_NO_STREAM_FLUSH> flag; in this case, data
-remaining in stream send buffers may not be transmitted to the peer. This flag
-may be used when a non-normal application condition has occurred and the
-delivery of data written to streams via L<SSL_write(3)> is no longer relevant.
+When used with a QUIC connection SSL object, SSL_shutdown() initiates a QUIC
+immediate close using QUIC B<CONNECTION_CLOSE> frames.
+
+SSL_shutdown_ex() may be used by an application to provide additional
+information to the peer on the reason why a connection is being shut down. The
+information which can be provided is as follows:
+
+=over 4
+
+=item I<quic_error_code>
+
+An optional 62-bit application error code to be signalled to the peer. The value
+must be in the range [0, 2**62-1], else the call to SSL_shutdown_ex() fails. If
+not provided, an error code of 0 is used by default.
+
+=item I<quic_reason>
+
+An optional zero-terminated (UTF-8) reason string to be signalled to the peer.
+The application is responsible for providing a valid UTF-8 string and OpenSSL
+will not validate the string. If a reason is not provided, or SSL_shutdown() is
+used, a zero-length string is used as the reason. If provided, the reason string
+is copied and stored inside the QUIC connection SSL object and need not remain
+allocated after the call to SSL_shutdown_ex() returns. Reason strings are
+bounded by the path MTU and may be silently truncated if they are too long to
+fit in a QUIC packet.
+
+Reason strings are intended for human diagnostic purposes only, and should not
+be used for application signalling.
+
+=back
+
+The arguments to SSL_shutdown_ex() are used only on the first call to
+SSL_shutdown_ex() (or SSL_shutdown()) for a given QUIC connection SSL object.
+These arguments are ignored on subsequent calls.
+
+=head2 Application Data Drainage Behaviour
+
+When using QUIC, SSL_shutdown() or SSL_shutdown_ex() ordinarily waits until all
+data written to a stream by an application has been acknowledged by the peer. In
+other words, the shutdown process waits until all data written by the
+application has been sent to the peer, and until the receipt of all such data is
+acknowledged by the peer. Only once this process is completed is thue shutdown
+considered complete.
+
+An exception to this is streams which terminated in a non-normal fashion, for
+example due to a stream reset; only streams which are non-terminated at the time
```

Comment:
What is a stream reset? Can this non-normal fashion be triggered by an application?

Reply:
Yes. See [RFC 9000 s. 2.4](https://www.rfc-editor.org/rfc/rfc9000.html#name-operations-on-streams), [RFC 9000 s. 19.4](https://www.rfc-editor.org/rfc/rfc9000.html#name-reset_stream-frames).

See SSL_stream_reset(3) for our API for this. A stream reset has application-defined semantics and implies an application doesn't care about any data which was written to the stream still being delivered. It is *somewhat* analogous to TCP RST for a single stream.

Reply:
Adding a reference to the `SSL_stream_reset(3)` page would be useful here.